onnxscript/rewriter/pattern.py

```diff
@@ -927,57 +1023,38 @@ def _apply_deltas(
             # TODO: simplify this
             last_deleted = deleted_nodes[-1]
             last_inserted = inserted_nodes[-1]
-
             assert len(last_deleted.outputs) == len(last_inserted.outputs)
-            del last_inserted.outputs[:]
-            for v in last_deleted.outputs:
-                v.node = last_inserted
-                last_inserted.outputs.append(v)
-
-            del nodes[i]
-
-            for new_node in reversed(inserted_nodes):
-                nodes.insert(i, new_node)
-                # bind the outputs to the graph
-                for output_name, value in zip(new_node.output_names, new_node.outputs):
-                    graph_or_function.values[output_name] = value
-            path_2 = True
-
-    assert not to_delete or not path_2, (
-        "Two different rules were applied. It will solved later. "
-        "Right now, the functions assumes all the changes come from one "
-        "rule."
-    )
-
-    if path_2:
-        for _, delta in deltas:
-            deleted_nodes, inserted_nodes = delta
-            inserted_input_output = []
-            for nd in inserted_nodes:
-                inserted_input_output += nd.inputs + nd.outputs
-            for old_node in deleted_nodes[0:-1]:
-                # Delete intermediary outputs from graph that are not used as
-                # outputs of the graph
-                for output in old_node.outputs:
-                    if not output.is_output and output not in inserted_input_output:
-                        graph_or_function.values.pop(output.name)
-                nodes.remove(old_node)
-
-    for i in to_delete:
-        position = existing_ids[i][0]
-        nodes[position] = None
-
-    for position, insert in sorted(to_insert.items(), reverse=True):
-        for v in reversed(insert):
-            nodes.insert(position, v)
-
-    position_to_delete = []
-    for i, n in enumerate(nodes):
-        if n is None:
-            position_to_delete.append(i)
-
-    for p in reversed(position_to_delete):
-        del nodes[p]
+            # Reconnect the users of the deleted node to use the new outputs
+            for last_deleted_output, last_inserted_output in zip(
+                last_deleted.outputs, last_inserted.outputs
+            ):
+                for node, index in last_deleted_output.consumers():
+                    node.replace_input_with(index, last_inserted_output)
```

Comment:
Great if this works. Just trying to understand what the IR allows us to do: at this point `last_inserted_output` is computed by a node that has not been inserted into the graph yet (I guess that happens below). Is that safe/permitted?

Reply:
Or, is it better to insert the nodes first, before updating the uses?

Reply:
Yes! The value-node connections are independent of the graph. This means we will need to maintain connections for them to be valid, but validity is easy to check

Reply:
> Or, is it better to insert the nodes first, before updating the consumers?

I think they are equivalent? Is there any pros to requiring this?

Reply:
Well, in one sense, it is a well-formed graph when we insert new nodes that compute new values. And then if we replace old-value with new-value in some node that consumers old-value. But if we replace old-value with new-value, where new-value is not yet a valid value in the graph, we have a temporarily invalid graph. That's why I was asking (since it could make a difference in some implementations, but I guess it does not in the current one.)